### ISSUE TITLE
Docker 18.09.1~3 breaks Ubuntu 18.04 DNS on the host 

### ISSUE THREAD

**Issue by vaskokj** [docker/libnetwork#2335]
Installing a fresh 18.04 with docker will break the hosts DNS resolution. My DNS server is a local server on the LAN. 

Steps to reproduce

Install a fresh 18.04 LTS
Install docker-ce

During kernel boot I see

`A start job is running for Docker socket for the API`

With a timer. The timer will count up to 3min and then will continue to boot.

Once the machine does boot DNS resolution is completely hosed on the machine. I can ping public and local systems by their IP, however no DNS resolution will work on the box. The only way I have successfully be able to get DNS resolution back functioning is to remove docker-ce.

I have tried this solution which sounds like the same issue however the solution does not work. 

https://superuser.com/questions/1336567/installing-docker-ce-in-ubuntu-18-04-breaks-internet-connectivity-of-host

I see other related issues, however they all seem to have to do with the docker containers not being able to reach the DNS servers. My issue is the host DNS gets broken. 
```

DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.1 LTS"
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
```

```
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 02:42:2d:08:cd:49 brd ff:ff:ff:ff:ff:ff
    inet 172.19.0.1/16 brd 172.19.255.255 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:2dff:fe08:cd49/64 scope link 
       valid_lft forever preferred_lft forever
```
```
2: enp0s25: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 48:4d:7e:f4:c2:a3 brd ff:ff:ff:ff:ff:ff
    inet 172.16.0.105/24 brd 172.16.0.255 scope global dynamic noprefixroute enp0s25
       valid_lft 84946sec preferred_lft 84946sec
    inet6 fe80::b01e:c394:974b:d59d/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever
```

```
docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; disabled; vendor preset: enabled)
   Active: active (running) since Fri 2019-02-01 17:45:58 CST; 43s ago
     Docs: https://docs.docker.com
 Main PID: 3779 (dockerd)
    Tasks: 58
   CGroup: /system.slice/docker.service
           ├─3779 /usr/bin/dockerd -H fd://
           ├─9128 setup-resolver /var/run/docker/netns/0892df9c72fb 127.0.0.11:52832 127.0.0.11:37891
           └─9145 /sbin/iptables --wait -L -n

eb 01 17:46:40 aiml5 dockerd[3779]: time="2019-02-01T17:46:40.653234806-06:00" level=info msg="No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4]"
Feb 01 17:46:40 aiml5 dockerd[3779]: time="2019-02-01T17:46:40.653282165-06:00" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844]"
Feb 01 17:46:40 aiml5 dockerd[3779]: time="2019-02-01T17:46:40.710380909-06:00" level=info msg="No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4]"
Feb 01 17:46:40 aiml5 dockerd[3779]: time="2019-02-01T17:46:40.710425668-06:00" level=info msg="IPv6 enabled; Adding default IPv6 external servers: [nameserver 2001:4860:4860::8888 nameserver 2001:4860:4860::8844]"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.462619558-06:00" level=error msg="stream copy error: reading from a closed fifo"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.462655436-06:00" level=error msg="stream copy error: reading from a closed fifo"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.491133668-06:00" level=warning msg="deleteServiceInfoFromCluster NetworkDB DeleteEntry failed for 94106a4c5a17ae1c7a310c324f59ca88f67a64b6823911130bbf86d3e762c9b8 wa8oec8otmy4ei0ri7x84b46u err:cannot delete entry endpoint_table with network id wa8oec8otmy4ei0ri7x84b46u and key 94106a4c5a17ae1c7a310c324f59ca88f67a64b6823911130bbf86d3e762c9b8 does not exist or is already being deleted"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.491188211-06:00" level=warning msg="rmServiceBinding deleteServiceInfoFromCluster perception_bag-player 94106a4c5a17ae1c7a310c324f59ca88f67a64b6823911130bbf86d3e762c9b8 aborted c.serviceBindings[skey] !ok"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.761852160-06:00" level=error msg="fcb7b234459f33ae73dac932a2d161431e445901ddd0c12e60262079f73009ec cleanup: failed to delete container from containerd: no such container"
Feb 01 17:46:41 aiml5 dockerd[3779]: time="2019-02-01T17:46:41.761896491-06:00" level=error msg="fatal task error" error="starting container failed: OCI runtime create failed: container_linux.go:344: starting container process caused \"process_linux.go:424: container init caused \\\"rootfs_linux.go:58: mounting \\\\\\\"/mnt/nfshome\\\\\\\" to rootfs \\\\\\\"/var/lib/docker/overlay2/3e1d57d6cb3838dcdc78024a7464c717b75dc39934340a887573621407083e31/merged\\\\\\\" at \\\\\\\"/var/lib/docker/overlay2/3e1d57d6cb3838dcdc78024a7464c717b75dc39934340a887573621407083e31/merged/mnt/nfshome\\\\\\\" caused \\\\\\\"no such file or directory\\\\\\\"\\\"\": unknown"

```

**Comment by vaskokj** [comment]
Did some more debugging on a different machine that was exhibiting the same exact issue.

The following is the journalctl -b output before and after where I am seeing one of the issues

> 
> Feb 06 11:35:18 aiml6 systemd[1]: Started Trigger anacron every hour.
> Feb 06 11:35:18 aiml6 systemd[1]: Starting Socket activation for snappy daemon.
> Feb 06 11:35:18 aiml6 systemd[1]: Started Daily Cleanup of Temporary Directories.
> Feb 06 11:35:18 aiml6 systemd[1]: Started Discard unused blocks once a week.
> Feb 06 11:35:18 aiml6 systemd[1]: Started ACPI Events Check.
> Feb 06 11:35:18 aiml6 systemd[1]: Started CUPS Scheduler.
> Feb 06 11:35:18 aiml6 systemd[1]: Reached target Paths.
> Feb 06 11:35:18 aiml6 systemd[1]: Listening on UUID daemon activation socket.
> Feb 06 11:35:18 aiml6 systemd[1]: Started Message of the Day.
> Feb 06 11:35:18 aiml6 systemd[1]: Started Daily apt download activities.
> Feb 06 11:35:18 aiml6 systemd[1]: Started Daily apt upgrade and clean activities.
> Feb 06 11:35:18 aiml6 systemd[1]: Reached target Timers.
> Feb 06 11:35:18 aiml6 systemd[1]: Listening on Socket activation for snappy daemon.
> **Feb 06 11:36:48 aiml6 systemd[1]: docker.socket: Starting timed out. Stopping.
> Feb 06 11:36:48 aiml6 systemd[977]: docker.socket: Failed to resolve group docker: Connection refused
> Feb 06 11:38:18 aiml6 systemd[1]: docker.socket: Stopping timed out (2). Killing.
> Feb 06 11:38:18 aiml6 systemd[1]: docker.socket: Failed with result 'timeout'.
> Feb 06 11:38:18 aiml6 systemd[1]: Failed to listen on Docker Socket for the API.
> Feb 06 11:38:18 aiml6 systemd[1]: Dependency failed for Docker Application Container Engine.
> Feb 06 11:38:18 aiml6 systemd[1]: docker.service: Job docker.service/start failed with result 'dependency'.**
> Feb 06 11:38:18 aiml6 systemd[1]: Reached target Sockets.
> Feb 06 11:38:18 aiml6 systemd[1]: Reached target Basic System.
> Feb 06 11:38:18 aiml6 systemd[1]: Started ACPI event daemon.
> Feb 06 11:38:18 aiml6 systemd[1]: Started Set the CPU Frequency Scaling governor.
> Feb 06 11:38:18 aiml6 systemd[1]: Starting LSB: Record successful boot for GRUB...
> Feb 06 11:38:18 aiml6 systemd[1]: Reached target Login Prompts.
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Restore /etc/resolv.conf if the system crashed before the ppp link was shut down...
> Feb 06 11:38:18 aiml6 acpid[987]: starting up with netlink and the input layer
> Feb 06 11:38:18 aiml6 acpid[987]: 8 rules loaded
> Feb 06 11:38:18 aiml6 acpid[987]: waiting for events: event logging is off
> Feb 06 11:38:18 aiml6 systemd[1]: Starting LSB: automatic crash report generation...
> Feb 06 11:38:18 aiml6 systemd[1]: Started Run anacron jobs.
> Feb 06 11:38:18 aiml6 anacron[1042]: Anacron 2.3 started on 2019-02-06
> Feb 06 11:38:18 aiml6 anacron[1042]: Normal exit (0 jobs run)
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Dispatcher daemon for systemd-networkd...
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Detect the available GPUs and deal with any system changes...
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Snappy daemon...
> Feb 06 11:38:18 aiml6 networkd-dispatcher[1050]: WARNING: systemd-networkd is not running, output will be incomplete.
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Avahi mDNS/DNS-SD Stack...
> Feb 06 11:38:18 aiml6 avahi-daemon[1093]: Found user 'avahi' (UID 116) and group 'avahi' (GID 122).
> Feb 06 11:38:18 aiml6 avahi-daemon[1093]: Successfully dropped root privileges.
> Feb 06 11:38:18 aiml6 avahi-daemon[1093]: avahi-daemon 0.7 starting up.
> Feb 06 11:38:18 aiml6 systemd[1]: Starting System Security Services Daemon...
> Feb 06 11:38:18 aiml6 snapd[1069]: AppArmor status: apparmor is enabled and all features are available
> Feb 06 11:38:18 aiml6 systemd[1]: Started CUPS Scheduler.
> Feb 06 11:38:18 aiml6 audit[1133]: AVC apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/etc/sssd/conf.d/" pid=1133 comm="sssd" requested_mask="r" denied_m
> Feb 06 11:38:18 aiml6 audit[1133]: AVC apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/usr/share/sssd/cfg_rules.ini" pid=1133 comm="sssd" requested_mask=
> Feb 06 11:38:18 aiml6 sssd[1133]: Starting up
> Feb 06 11:38:18 aiml6 systemd[1]: Starting LSB: Speech Dispatcher...
> Feb 06 11:38:18 aiml6 kernel: kauditd_printk_skb: 30 callbacks suppressed
> Feb 06 11:38:18 aiml6 kernel: audit: type=1400 audit(1549474698.863:42): apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/etc/sssd/conf.d/" pid=1133 comm=
> Feb 06 11:38:18 aiml6 kernel: audit: type=1400 audit(1549474698.863:43): apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/usr/share/sssd/cfg_rules.ini" pi
> Feb 06 11:38:18 aiml6 sssd[be[1159]: Starting up
> Feb 06 11:38:18 aiml6 audit[1159]: AVC apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/etc/krb5.conf.d/" pid=1159 comm="sssd_be" requested_mask="r" denie
> Feb 06 11:38:18 aiml6 audit[1159]: AVC apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/etc/gss/mech.d/" pid=1159 comm="sssd_be" requested_mask="r" denied
> Feb 06 11:38:18 aiml6 systemd[1]: Starting Modem Manager...
> Feb 06 11:38:18 aiml6 kernel: audit: type=1400 audit(1549474698.899:44): apparmor="ALLOWED" operation="open" profile="/usr/sbin/sssd" name="/etc/krb5.conf.d/" pid=1159 comm=

**Comment by vaskokj** [comment]
So did some _more_ debugging and think I found the issue, and workaround (with the help of "ada" and "programmerq" on IRC). 

To make a long story short, I'm using FreeIPA for my login service on my linux machines, which means I need DNS resolution to authenticate to the IPA server. However, my docker.socket is starting before DNS resolution is up and when docker.socket tries to resolve the docker "group" from IPA, it fails to resolve since its looking at the FreeIPA server for that information and without DNS resolution it fails to resolve, which stops docker.socket from starting up properly. 

The key was the following error message

> docker.socket - Docker Socket for the API
>    Loaded: loaded (/lib/systemd/system/docker.socket; enabled; vendor preset: enabled)
>    Active: failed (Result: timeout) since Wed 2019-02-06 14:12:30 CST; 4min 53s ago
>    Listen: /var/run/docker.sock (Stream)
> 
> Feb 06 14:09:30 aiml6 systemd[1]: Starting Docker Socket for the API.
> Feb 06 14:11:00 aiml6 systemd[1]: docker.socket: Starting timed out. Stopping.
> **Feb 06 14:11:00 aiml6 systemd[985]: docker.socket: Failed to resolve group docker: Connection refused**
> Feb 06 14:12:30 aiml6 systemd[1]: docker.socket: Stopping timed out (2). Killing.
> Feb 06 14:12:30 aiml6 systemd[1]: docker.socket: Failed with result 'timeout'.
> Feb 06 14:12:30 aiml6 systemd[1]: Failed to listen on Docker Socket for the API.

This does NOT happen on Ubuntu 16.04 with Docker 18.09 since it appears that docker does NOT include a docker.socket systemd service file in Ubuntu 16.04 LTS. 

A workaround that seems to work:

1) add a local "docker" group so the docker.socket service can find it locally and doesn't have to reach out to the IPA server
2) disable docker.service and docker.socket services in systemd 
`sudo systemctl disable docker.service`
`sudo systemctl disable docker.socket`

Since making change 1) I have yet to see this issue crop up again. 

I feel a proper fix would be to have docker.socket start up _AFTER_ the DNS resolution services come up (systemd-resolved). I have tried adding `After=` in the docker.socket but was getting all types of error messages doing that. I am not familiar enough with systemd to know how to resolve this unfortunately.
